train.py

```diff
@@ -28,17 +34,54 @@
 )
 from torchtitan.profiling import maybe_enable_memory_snapshot, maybe_enable_profiling
 
+try:
+    from torch.distributed.tensor.experimental import context_parallel
+except ImportError:
+    print(
+        f"PyTorch version {torch.__version__} does not include the experimental "
+        "Context Parallel API. Please update to a newer version."
+    )
+
+
+def get_train_context(
+    enable_loss_parallel: bool,
+    enable_compiled_autograd: bool,
+    cp_mesh: Optional[DeviceMesh] = None,
+):
+    if cp_mesh is not None:
+        context_parallel_ctx = partial(context_parallel, mesh=cp_mesh)
 
-def get_train_context(enable_loss_parallel: bool, enable_compiled_autograd: bool):
     @contextlib.contextmanager
-    def context():
+    def context(
+        cp_buffers: List[torch.Tensor],
+        cp_seq_dims: List[int],
+        cp_no_restore_buffers: Set[torch.Tensor],
```

Comment:
Instead of passing CP related args to the general `context()`, we should create a "context parallel" context as the arg. And later on call `stack.enter_context()` to enter this context. We can create a separate function to generate this context given the CP related args. This function should be put into another file, e.g.`torchtitan/parallelisms/utils.py`.

Reply:
I'm not sure if creating CP context on demand is a good idea because of the overhead. Besides, the computation in ring attention is very regular and repeated so re-using the context looks reasonable to me. cc @fegin